crates/uv-interpreter/src/python_query.rs

```diff
@@ -144,130 +73,404 @@ pub fn find_requested_python(
-            .or_else(|_| which::which("python"))
-            .map_err(|_| Error::NoPythonInstalledUnix)?
-    } else if cfg!(windows) {
-        // TODO(konstin): Is that the right order, or should we look for `py --list-paths` first? With the current way
-        // it works even if the python launcher is not installed.
-        if let Ok(python) = which::which_in(
-            "python.exe",
-            env::var_os("UV_TEST_PYTHON_PATH"),
-            current_dir,
-        )
-        .or_else(|_| which::which("python.exe"))
-        {
-            python
+    find_python(PythonVersionSelector::Default, platform, cache)?.ok_or(Error::NoPythonInstalled)
+}
+
+/// Finds a python version matching `selector`.
+/// It searches for an existing installation in the following order:
+/// * (windows): Discover installations using `py --list-paths` (PEP514). Continue if `py` is not installed.
+/// * Search for the python binary in `PATH` (or `UV_TEST_PYTHON_PATH` if set). Visits each path and for each path resolves the
+///   files in the following order:
+///   * Major.Minor.Patch: `pythonx.y.z`, `pythonx.y`, `python.x`, `python`
+///   * Major.Minor: `pythonx.y`, `pythonx`, `python`
+///   * Major: `pythonx`, `python`
+///   * Default: `python3`, `python`
+///   * (windows): For each of the above, test for the existence of `python.bat` shim (pyenv-windows) last.
+///
+/// (Windows): Filter out the windows store shim (Enabled in Settings/Apps/Advanced app settings/App execution aliases).
+fn find_python(
+    selector: PythonVersionSelector,
+    platform: &Platform,
+    cache: &Cache,
+) -> Result<Option<Interpreter>, Error> {
+    #[allow(non_snake_case)]
+    let UV_TEST_PYTHON_PATH = env::var_os("UV_TEST_PYTHON_PATH");
+
+    if UV_TEST_PYTHON_PATH.is_none() {
+        // Use `py` to find the python installation on the system.
+        #[cfg(windows)]
+        match windows::py_list_paths(selector, platform, cache) {
+            Ok(Some(interpreter)) => return Ok(Some(interpreter)),
+            Ok(None) => {
+                // No matching Python version found, continue searching PATH
+            }
+            Err(Error::PyList(error)) => {
+                if error.kind() == std::io::ErrorKind::NotFound {
+                    tracing::warn!(
+                        "`py` is not installed. Falling back to searching Python on the path"
+                    );
+                    // Continue searching for python installations on the path.
+                }
+            }
+            Err(error) => return Err(error),
+        }
+    }
+
+    let possible_names = selector.possible_names();
+    let mut checked_installs = HashSet::new();
+
+    #[allow(non_snake_case)]
+    let PATH = UV_TEST_PYTHON_PATH
+        .or(env::var_os("PATH"))
+        .unwrap_or_default();
+
+    for path in env::split_paths(&PATH) {
+        for name in possible_names.iter().flatten() {
+            if let Ok(paths) = which::which_in_global(&**name, Some(&path)) {
```

Comment:
Can you add a comment explaining why we're using `which` here instead of joining ourselves